File: implement-trie-prefix-tree/sangbeenmoon.java
import java.util.HashMap;
import java.util.Map;

class Trie {

    Map<Character, Trie> map;
    boolean isEnd = false;

    public Trie() {
        map = new HashMap<>();
    }

    public void insert(String word) {
        Trie trie;
        if (map.containsKey(word.charAt(0))) {
            trie = map.get(word.charAt(0));
        } else {
            trie = new Trie();
        }
        map.put(word.charAt(0), trie);

        if (word.length() == 1) {
            map.get(word.charAt(0)).isEnd = true;
            return;
        }

        trie.insert(word.substring(1));
    }

    public boolean search(String word) {
        if (map.containsKey(word.charAt(0))) {
            if (word.length() == 1) {
                return map.get(word.charAt(0)).isEnd;
            }
            return map.get(word.charAt(0)).search(word.substring(1));
        }
        return false;
    }

    public boolean startsWith(String prefix) {
        if (map.containsKey(prefix.charAt(0))) {
            if (prefix.length() == 1) {
                return true;
            }
            return map.get(prefix.charAt(0)).startsWith(prefix.substring(1));
        }
        return false;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: implement-trie-prefix-tree/sangbeenmoon.py
class Trie:

    children = {}
    is_end = False

    def __init__(self):
        self.children = {}
        self.is_end = False
        return

    def insert(self, word: str, i=0) -> None:

        if i == len(word):
            self.is_end = True
            return

        target = word[i]

        if target in self.children:
            node = self.children[target]
            node.insert(word, i+1)
        else:
            node = Trie()
            node.insert(word, i+1)
            self.children[target] = node
        
    def search(self, word: str, i=0) -> bool:

        if i == len(word):
            return self.is_end

        target = word[i]

        if target in self.children:
            node = self.children[target]
            return node.search(word, i+1)
        else:
            return False
        

    def startsWith(self, prefix: str, i=0) -> bool:
        if i == len(prefix):
            return True

        target = prefix[i]

        if target in self.children:
            node = self.children[target]
            return node.startsWith(prefix, i+1)
        else:
            return False        


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)
